httpbp/server.go

```diff
@@ -0,0 +1,271 @@
+		TrustHandler:    args.TrustHandler,
+		EdgeContextImpl: args.Baseplate.EdgeContextImpl(),
+	})
+	wrappers = append(wrappers, args.Middlewares...)
+
+	factory := httpHandlerFactory{middlewares: wrappers}
+	for pattern, endpoint := range args.Endpoints {
+		args.EndpointRegistry.Handle(
+			string(pattern),
+			factory.NewHandler(endpoint.Name, endpoint.Handle, endpoint.Middlewares...),
+		)
+	}
+	return args, nil
+}
+
+// NewBaseplateServer returns a new HTTP implementation of a Baseplate
+// server with the given ServerArgs.
+//
+// The Endpoints given in the ServerArgs will be wrapped using the
+// default Baseplate Middleware as well as any additional Middleware
+// passed in.
+func NewBaseplateServer(args ServerArgs) (baseplate.Server, error) {
+	args, err := args.SetupEndpoints()
+	if err != nil {
+		return nil, err
+	}
+
+	srv := &http.Server{
+		Addr:         args.Baseplate.Config().Addr,
+		Handler:      args.EndpointRegistry,
+		ReadTimeout:  args.Baseplate.Config().Timeout,
+		WriteTimeout: args.Baseplate.Config().Timeout,
+	}
+	for _, f := range args.OnShutdown {
+		srv.RegisterOnShutdown(f)
+	}
+	return &server{args.Baseplate, srv}, nil
+}
+
+type server struct {
+	bp  baseplate.Baseplate
+	srv *http.Server
+}
+
+func (s server) Baseplate() baseplate.Baseplate {
+	return s.bp
+}
+
+func (s server) Serve() error {
+	// ListenAndServe always returns a non-nil error, http.ErrServerClosed is the
+	// "expected" error for it to return after being shutdown.
+	//
+	// https://golang.org/pkg/net/http/#Server.ListenAndServe
+	err := s.srv.ListenAndServe()
+	if errors.Is(err, http.ErrServerClosed) {
+		return nil
+	}
+
+	return err
+}
+
+func (s server) Stop() error {
+	return s.srv.Shutdown(context.Background())
+}
+
+// NewTestBaseplateServer returns a new HTTP implementation of a Baseplate
+// server with the given ServerArgs that uses a Server from httptest rather than
+// a real server.
+//
+// The underlying httptest.Server is started when the the test BaseplateServer
+// is created and does not need to be started manually.
+// It is closed by calling Stop, Stop should not be called more than once.
+// Serve does not need to be called but will wait until Stop is called to exit
+// if it is called.
+func NewTestBaseplateServer(args ServerArgs) (baseplate.Server, *httptest.Server, error) {
+	args, err := args.SetupEndpoints()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	ts := httptest.NewServer(args.EndpointRegistry)
+	return &testServer{
+		bp:         args.Baseplate,
+		onShutdown: args.OnShutdown,
+		srv:        ts,
+		wg:         wg,
+	}, ts, nil
+}
+
+type testServer struct {
+	bp         baseplate.Baseplate
+	onShutdown []func()
+	srv        *httptest.Server
+	wg         *sync.WaitGroup
+}
+
+func (s *testServer) Baseplate() baseplate.Baseplate {
+	return s.bp
+}
+
+func (s *testServer) Serve() error {
+	s.wg.Wait()
+	return nil
+}
+
+func (s *testServer) Stop() error {
+	s.srv.Close()
```

Comment:
return the error of `Close`?

Reply:
It doesn't return one:

https://golang.org/pkg/net/http/httptest/#Server.Close